app/src/main/java/protect/card_locker/ScanActivity.java

```diff
@@ -237,11 +245,41 @@ private void showCameraPermissionMissingText(boolean show) {
 
     }
 
+    private void scaleScreen() {
+        DisplayMetrics metrics = getResources().getDisplayMetrics();
+        float ratio = (float) metrics.heightPixels / metrics.widthPixels;
+        boolean shouldScaleSmaller = ratio <= 0.8f;
+        boolean shouldScaleMedium = ratio <= 1.1f && !shouldScaleSmaller;
+        if (shouldScaleMedium) {
+            customBarcodeScannerBinding.cameraPermissionDeniedLayout.cameraPermissionDeniedIcon.setVisibility( View.GONE );
+        } else {
+            int buttonMinHeight = getResources().getDimensionPixelSize(R.dimen.scan_button_min_height);
+            customBarcodeScannerBinding.cameraPermissionDeniedLayout.cameraPermissionDeniedTitle.setVisibility(shouldScaleSmaller ? View.GONE : View.VISIBLE);
+            customBarcodeScannerBinding.cameraPermissionDeniedLayout.cameraPermissionDeniedIcon.setVisibility(shouldScaleSmaller ? View.GONE : View.VISIBLE);
+            customBarcodeScannerBinding.cameraPermissionDeniedLayout.cameraPermissionDeniedMessage.setTextSize(TypedValue.COMPLEX_UNIT_PX, getResources().getDimensionPixelSize(shouldScaleSmaller ? R.dimen.no_data_min_textSize : R.dimen.no_data_max_textSize));
+            customBarcodeScannerBinding.addFromImage.setTextSize(TypedValue.COMPLEX_UNIT_PX, getResources().getDimensionPixelSize(shouldScaleSmaller ? R.dimen.scan_button_min_textSize : R.dimen.scan_button_max_textSize));
+            customBarcodeScannerBinding.addManually.setTextSize(TypedValue.COMPLEX_UNIT_PX, getResources().getDimensionPixelSize(shouldScaleSmaller ? R.dimen.scan_button_min_textSize : R.dimen.scan_button_max_textSize));
+            customBarcodeScannerBinding.addFromImage.setMinimumHeight(shouldScaleSmaller ? getResources().getDimensionPixelSize(R.dimen.scan_button_min_height) : buttonDefaultMinHeight);
+            customBarcodeScannerBinding.addFromImage.setMinimumWidth(shouldScaleSmaller ? buttonMinHeight : buttonDefaultMinHeight);
+            customBarcodeScannerBinding.addManually.setMinimumHeight(shouldScaleSmaller ? buttonMinHeight : buttonDefaultMinHeight);
+            customBarcodeScannerBinding.addManually.setMinimumWidth(shouldScaleSmaller ? buttonMinHeight : buttonDefaultMinHeight);
+            customBarcodeScannerBinding.addFromImage.setMinHeight(shouldScaleSmaller ? buttonMinHeight : buttonDefaultMinHeight);
+            customBarcodeScannerBinding.addFromImage.setMinWidth(shouldScaleSmaller ? buttonMinHeight : buttonDefaultMinHeight);
+            customBarcodeScannerBinding.addManually.setMinHeight(shouldScaleSmaller ? buttonMinHeight : buttonDefaultMinHeight);
+            customBarcodeScannerBinding.addManually.setMinWidth(shouldScaleSmaller ? buttonMinHeight : buttonDefaultMinHeight);
+        }
+    }
+
+    private void saveDefaultUIValues() {
+        buttonDefaultMinHeight = customBarcodeScannerBinding.addFromImage.getMinHeight();
+        buttonDefaultMaxHeight = customBarcodeScannerBinding.addFromImage.getMinWidth();
+    }
```

Comment:
Originally I was going to suggest moving the saving the default UI values into the start of `scaleScreen`, but looking at it more I realize I'm not quite sure what all the `setMin*` is for. As I understand Android UI (which, admittedly, I don't really do well) the fact that the buttons use `wrap_content` for both width and height means they should scale to whatever the font size is.

I also tried this on my Fairphone 3 and the buttons still overlayed the text, so I guess the scaling isn't quite right. On top of that I tried removing all the `setMin*` and noticed no difference at all, not even when comparing the screenshots before and after my change by rapidly alt-tabbing between them.

![image](https://user-images.githubusercontent.com/1885159/196542210-753a8d39-e1f5-4965-8c60-8ac87bea1667.png)

---

I have the feeling this way of resizing may simply not be the correct way to go about it. Even when tweaking the numbers, we may still run into unexpected cases.

I think it would possibly be cleaner to just tell Android the buttons at the bottom should keep their regular space and the help text at the top should be squashed to fit in the rest of the space, possibly using `setAutoSizeTextTypeUniformWithConfiguration` to scale. But again, UI is one of my weakest points, especially in the way Android XML works. I'm hoping some Android community could perhaps help us figure out the best way to go about this.

Reply:
> Originally I was going to suggest moving the saving the default UI values into the start of `scaleScreen`, but looking at it more I realize I'm not quite sure what all the `setMin*` is for. As I understand Android UI (which, admittedly, I don't really do well) the fact that the buttons use `wrap_content` for both width and height means they should scale to whatever the font size is.
 Yes, buttons will be scaled based on the size of the text, however, there's a minimum size for buttons to be scaled. When I set the button text size to smaller, the buttons' sizes won't change because they have already reached the minimum size. To solve this problem, I have to override the minimum size of those buttons to scale them down even further.

> I also tried this on my Fairphone 3 and the buttons still overlayed the text, so I guess the scaling isn't quite right. On top of that I tried removing all the setMin* and noticed no difference at all, not even when comparing the screenshots before and after my change by rapidly alt-tabbing between them.
I think maybe it is not reliable to set a specific screen ratio for the UI to scale, I'd better find another way to do it.

> I think it would possibly be cleaner to just tell Android the buttons at the bottom should keep their regular space and the help text at the top should be squashed to fit in the rest of the space, possibly using setAutoSizeTextTypeUniformWithConfiguration to scale. But again, UI is one of my weakest points, especially in the way Android XML works. I'm hoping some Android community could perhaps help us figure out the best way to go about this.

Thank you for the advice and I will look into this to try to find a better way to scale it.

Reply:
After having tried to solve the problem with ` TextViewCompat.setAutoSizeTextTypeUniformWithConfiguration() `, I found out this function may only be meant for scaling texts based on the font size setting in Android system settings.

When I was poking around with Google's documents about support for different sizes of screens([the link is here](https://developer.android.com/guide/topics/large-screens/support-different-screen-sizes)), I realized that it might be a good idea to use screen's height as the scale factor other than ratio, this is because for most of the phones, whether the content can be vertically contained or not is mainly decided by the height of the screen. I think using height as the scale factor might be more universally compatible with most phones. I have run it on my Android emulator with different screen sizes and it looked fine.

I made a new commit [b44d998](https://github.com/CatimaLoyalty/Android/pull/1095/commits/b44d99822dfa783caf26838cdf9ea69f9ad6c803) to push the code.